packages/flutter_tools/lib/src/test/integration_test_device.dart

```diff
@@ -97,7 +97,14 @@ class IntegrationTestTestDevice implements TestDevice {
       );
     });
 
-    unawaited(remoteMessages.pipe(controller.local.sink));
+    remoteMessages.listen(
+      (String s) => controller.local.sink.add(s),
+      onError: (Object error, StackTrace stack) => controller.local.sink.addError(error, stack),
+    );
+    unawaited(vmService.service.onDone.whenComplete(
+      () => controller.local.sink.close(),
+    ));
```

Comment:
nit: Can this be merged into the `remoteMessages.listen` above?

```dart
remoteMessages.listen(
  (String s) => controller.local.sink.add(s),
  onError: (Object error, StackTrace stack) => controller.local.sink.addError(error, stack),
  onDone: () => controller.local.sink.close(),
);
```

Reply:
The `remoteMessages` stream is not closed when the on-device process crashes.

That event appears to be signaled only by the VM service's `onDone`.